### ISSUE TITLE
model.update_run_params not reliable the in the way I use it.

### ISSUE THREAD

**Issue by russelljjarvis** [russelljjarvis/neuronunit#11]
Hi @rgerkin
I have tracked down the source of the bug that I spoke about today. Its very subtle, and it may take a long time to fix.

Its because the method I created update_run_params does not reliably update HOC/NEURON parameters. So the real attributes that the NEURON model has changes in a CPU dependent manner in the parallel rheobase search. That means some CPUs are searching rheobase for one set of model parameters, while other cpus are searching rheobase on another set of model parameters. The stack trace below is hard to follow but it captures, the phenemona:

```
5 this is cnt, exhausted cnt? 
False
[<__main__.VirtualModel object at 0x7f45410f2630>, <__main__.VirtualModel object at 0x7f45410f24e0>, <__main__.VirtualModel object at 0x7f45410f2860>, <__main__.VirtualModel object at 0x7f45410f2400>, <__main__.VirtualModel object at 0x7f45410f2320>, <__main__.VirtualModel object at 0x7f45410f25c0>, <__main__.VirtualModel object at 0x7f45455cef98>, <__main__.VirtualModel object at 0x7f45410f23c8>, <__main__.VirtualModel object at 0x7f45410f2ac8>, <__main__.VirtualModel object at 0x7f45410f2940>, <__main__.VirtualModel object at 0x7f45410f2978>, <__main__.VirtualModel object at 0x7f45410f29e8>, <__main__.VirtualModel object at 0x7f45410f28d0>, <__main__.VirtualModel object at 0x7f45410f2a20>, <__main__.VirtualModel object at 0x7f45410f2fd0>, <__main__.VirtualModel object at 0x7f45410f2f98>]
False
False
False
RS_pop[0] { nseg=1  L=10  Ra=35.4
	/*location 0 attached to cell 0*/
	/* First segment only */
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
	insert RS { v0=-60 k=0.0007 vr=-60 vt=-40 vpeak=35 a=0.03 b=-4.25e-09 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.0648148}
}
1.0
False
False
RS_pop[0] { nseg=1  L=10  Ra=35.4
	/*location 0 attached to cell 0*/
	/* First segment only */
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
False
False
	insert RS { v0=-60 k=0.0007 vr=-60 vt=-40 vpeak=35 a=0.03 b=-4.25e-09 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.0617284}
}
1.0
RS_pop[0] { nseg=1  L=10  Ra=35.4
	/*location 0 attached to cell 0*/
	/* First segment only */
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
	insert RS { v0=-60 k=0.0007 vr=-62.5 vt=-40 vpeak=35 a=0.03 b=-0.002 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.0648148}
}
1.0
False
False
False
False
RS_pop[0] { nseg=1  L=10  Ra=35.4
	/*location 0 attached to cell 0*/
	/* First segment only */
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
	insert RS { v0=-60 k=0.0007 vr=-60 vt=-40 vpeak=35 a=0.03 b=-0.002 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.062963}
}
RS_pop[0] { nseg=1  L=10  Ra=35.4
	/*location 0 attached to cell 0*/
	/* First segment only */
1.0
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
	insert RS { v0=-60 k=0.0007 vr=-60 vt=-40 vpeak=35 a=0.03 b=-0.002 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.062963}
}
1.0
False
False
RS_pop[0] { nseg=1  L=10  Ra=35.4
	/*location 0 attached to cell 0*/
	/* First segment only */
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
	insert RS { v0=-60 k=0.0007 vr=-60 vt=-40 vpeak=35 a=0.03 b=-4.25e-09 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.0635802}
}
1.0
False
False
RS_pop[0] { nseg=1  L=10  Ra=35.4
	/*location 0 attached to cell 0*/
	/* First segment only */
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
	insert RS { v0=-60 k=0.0007 vr=-62.5 vt=-40 vpeak=35 a=0.03 b=-0.002 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.0617284}
}
1.0
False
False
RS_pop[0] { nseg=1  L=10  Ra=35.4
	/*location 0 attached to cell 0*/
	/* First segment only */
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
	insert RS { v0=-60 k=0.0007 vr=-60 vt=-40 vpeak=35 a=0.03 b=-0.002 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.062963}
}
1.0
False
False
RS_pop[0] { nseg=1  L=10  Ra=35.4
	/*location 0 attached to cell 0*/
	/* First segment only */
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
	insert RS { v0=-60 k=0.0007 vr=-60 vt=-40 vpeak=35 a=0.03 b=-4.25e-09 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.0635802}
}
1.0
False
False
RS_pop[0] { nseg=1  L=10  Ra=35.4
	/*location 0 attached to cell 0*/
	/* First segment only */
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
	insert RS { v0=-60 k=0.0007 vr=-60 vt=-40 vpeak=35 a=0.03 b=-4.25e-09 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.0648148}
}
1.0
False
False
RS_pop[0] { nseg=1  L=10  Ra=35.4
	/*location 0 attached to cell 0*/
	/* First segment only */
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
	insert RS { v0=-60 k=0.0007 vr=-62.5 vt=-40 vpeak=35 a=0.03 b=-0.002 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.0617284}
}
False
1.0
False
RS_pop[0] { nseg=1  L=10  Ra=35.4
	/*location 0 attached to cell 0*/
	/* First segment only */
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
	insert RS { v0=-60 k=0.0007 vr=-60 vt=-40 vpeak=35 a=0.03 b=-4.25e-09 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.0617284}
}
False
1.0
False
False
False
RS_pop[0] { nseg=1  L=10  Ra=35.4
	/*location 0 attached to cell 0*/
	/* First segment only */
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
	insert RS { v0=-60 k=0.0007 vr=-60 vt=-40 vpeak=35 a=0.03 b=-0.002 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.062963}
}
RS_pop[0] { nseg=1  L=10  Ra=35.4
1.0
	/*location 0 attached to cell 0*/
	/* First segment only */
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
	insert RS { v0=-60 k=0.0007 vr=-60 vt=-40 vpeak=35 a=0.03 b=-0.002 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.062963}
}
1.0
False
False
RS_pop[0] { nseg=1  L=10  Ra=35.4
	/*location 0 attached to cell 0*/
	/* First segment only */
	insert morphology { diam=10}
	insert capacitance { cm=31.831}
	insert RS { v0=-60 k=0.0007 vr=-60 vt=-40 vpeak=35 a=0.03 b=-4.25e-09 c=-50 d=0.1 C=0.0001}
	insert RS_Iext { delay=100 duration=1000 amplitude=0.0648148}
}
```

**Comment by rgerkin** [comment]
@russelljjarvis 
I propose that you decide how much faster you think things will be in the long run with a good parallel rheobase search versus the usual serial binary rheobase search inside a larger parallel model parameter search, and then use that to determine how much effort it is worth spending to fix the bug.  
If you think you can better allocate your time to other parts of this project, we can shut down parallel rheobase search and maybe revisit it another day.  But you can work on it now if you think it will be productive.

**Comment by russelljjarvis** [comment]
I believe I have fixed a lot of these problems currently in the latest version of [grid search ](https://github.com/russelljjarvis/neuronunit/blob/dev/neuronunit/tests/grid_search.py)

![image](https://cloud.githubusercontent.com/assets/7786645/24583986/00ccdd5c-1712-11e7-8fe7-f8c159f89daa.png)

Some of the reason for the problem was also I was just doing: 
```
model.inject_square_current(current)
    n_spikes = model.get_spike_count()
```

However for some tests n_spikes is not 1 but 0, so `get_spike_count` is not a reliable sanity check for score progress either.

The approach below seems to work.
```
sane = get_neab.suite.tests[3].sanity_check(value*pq.pA,model)

    uc = {'amplitude':value}
    current = params.copy()['injected_square_current']
    current.update(uc)
    current = {'injected_square_current':current}
    if len(model.attrs) == 0:
        model.update_run_params(vm.attrs)
    model.inject_square_current(current)
    n_spikes = model.get_spike_count()
    print(n_spikes)
    assert n_spikes == 1
    if sane == True and n_spikes == 1:
        for i in [4,5,6]:
            get_neab.suite.tests[i].params['injected_square_current']['amplitude']=value*pq.pA
        get_neab.suite.tests[0].prediction={}
        score = get_neab.suite.tests[0].prediction['value']=value*pq.pA
        score = get_neab.suite.judge(model)#passing in model, changes model
        import neuronunit.capabilities as cap
        spikes_numbers=[]
        plt.clf()
        plt.hold(True)
        for k,v in score.related_data.items():
            spikes_numbers.append(cap.spike_functions.get_spike_train(((v.values[0]['vm']))))
            plt.plot(model.results['t'],v.values[0]['vm'])
        plt.savefig(str(model.name)+'.png')
        plt.clf()
```

**Comment by russelljjarvis** [comment]
The Rheobase test as used by the GA is now looking sane also (see graphs below).

nsga is now using the same parallel rheobase method as grid search. I have found that its okay to use methods defined in modules, as long as they are not bound to, or nested inside an object.

This might mean that we can use methods defined in the tests/__init__.py class if we can unbind them from test object and then equate them to a new method label at run level 0.

![image](https://cloud.githubusercontent.com/assets/7786645/24629092/b148a836-186c-11e7-94ae-5d825cf59322.png)
![image](https://cloud.githubusercontent.com/assets/7786645/24629096/b4dbafa2-186c-11e7-9734-c1320eec39d9.png)
![image](https://cloud.githubusercontent.com/assets/7786645/24629101/b853a860-186c-11e7-8d46-0650f924ae4b.png)
![image](https://cloud.githubusercontent.com/assets/7786645/24629109/bafaee34-186c-11e7-8ffc-a2ec0a1e49f1.png)
![image](https://cloud.githubusercontent.com/assets/7786645/24629114/be4a3680-186c-11e7-85c3-a7916e27e5b5.png)
![image](https://cloud.githubusercontent.com/assets/7786645/24629119/c39250c8-186c-11e7-8788-7346583f6df7.png)

![image](https://cloud.githubusercontent.com/assets/7786645/24629123/c6788f8c-186c-11e7-948f-5c37e1062172.png)

**Comment by rgerkin** [comment]
In Python 3 you can do something like this:  
```
class A:
  def f(self,x):
    return x**2

a = A()
b = a.f.__func__
b(a,3) # prints 9
```
So `__func__` turns the instance method `a.f` into a regular function `b` which, instead of just taking the argument `x`, also takes an instance, i.e. `a` itself.  So maybe you can use `func` to do things with methods defined inside of classes, provided you already have instances of those classes in memory.

**Comment by russelljjarvis** [comment]
I will try this later. My priority is shifting to writing the abstract, poster, and seminar about this work.

I think it will be useful if I can plot the observations versus the best candidate solution, and the worst candidate solution for the presentation.

Most scientists understand the notion of curve fitting (because of regression in pearson correlation coefficients and or because the sum of least squares algorithm), and they probably can understand how you can get an intuitive sense of what would give a high error versus a low error, if I show them graphically the observation plotted against the prediction. And accompany this illustration with a qualitative explaination of what was tested.

**Comment by rgerkin** [comment]
I agree.
